### ISSUE TITLE
`torch.autograd.Function` subclasses *sometimes* throw away custom subclasses

### ISSUE THREAD

**Issue by nelhage** [pytorch/pytorch#58212]
## 🐛 Bug

`apply`ing a torch.autograd.Function usually returns the same `torch.Tensor` subclass as the input, but not always. This seems to happen when an input `requires_grad` and potentially some other requirements.

## To Reproduce

```python
import torch
import torch.autograd

class IdFunction(torch.autograd.Function):
  @staticmethod
  def forward(ctx, arg):
    return arg

  @staticmethod
  def backward(ctx, arg):
    return arg

class MyTensor(torch.Tensor):
  pass

x = torch.ones([1, 5, 5]).as_subclass(MyTensor)
x.requires_grad_()
applied = IdFunction.apply(x)
assert type(applied) is MyTensor, f"Expected MyTensor, got {type(applied)}"
```

If you remove the `x.requires_grad_()` call, the assertion passes.

## Expected behavior

I expect `torch.autograd.Function` to preserve the subclass passed in.
Barring that, I expect it to behave consistently about whether it preserves it or throws it away.

The documentation claims

> As of version 1.7.0, methods on torch.Tensor and functions in public torch.* namespaces applied on torch.Tensor subclasses will return subclass instances instead of torch.Tensor instances:

It's a bit unclear to me if that's meant to encompass custom `torch.autograd.Function` implementations that are built on such primitives.

## Environment

PyTorch version: 1.8.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: `13.0.0-++20210418052640+d480f968ad8b-1~exp1~20210418153358.383`
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 2080 SUPER
Nvidia driver version: 450.119.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.8.1
[pip3] torchvision==0.9.1
[conda] Could not collect

**Comment by ezyang** [comment]
Hi @nelhage ;) This is pretty similar to #47117 but there's no `save_for_backward` involved here. This might actually just be straight up fixed by #56017, if you are able to build PyTorch from source give that version a try.

**Comment by nelhage** [comment]
Hi @ezyang :)

I built from source on #56017 and it doesn't seem to have resolved the problem. Let me know if there are other tests I can run.

I noticed #47117 when searching for this and was hoping the lack of `save_for_backward` made this one more minimal or straightforward, but then forgot to flag the related bug in the report!

**Comment by albanD** [comment]
Hi,

What happens here is that we actually need the output of the custom Function to be a different object from the input.
So we internally do a shallow copy if you return the input as-is. In your case, that would create a plain Tensor and loos the subclass.

If you make sure that you never return the input, then the subclass should be maintained on exit.

**Comment by nelhage** [comment]
In that case, I'll back up and explain how I got here. I found this behavior while trying to understand the failure of the following program (itself minimized from my original code):

```python
import torch
from torch import nn

class MyTensor(torch.Tensor):
  pass

class Model(nn.Module):
  def __init__(self):
    super().__init__()

  def forward(self, input):
    assert isinstance(input, MyTensor)
    return input ** 2


m = Model()
t = torch.ones([1, 5, 5]).as_subclass(MyTensor)
t.requires_grad_()
m(t)

handle = m.register_full_backward_hook(lambda *args: None)

# The assert in `forward` will trigger, but only for this invocation, because of the hook
m(t)
```

The observed behavior is that the `assert isinstance(input, MyTensor)` check fails only on the *second* invocation of `m(t)`, with the backwards hook registered. The reason is that full backwards hooks actually call [this function](https://github.com/pytorch/pytorch/blob/4f28c0b5909d5122b0beffa49a579fc0d5fe0f80/torch/nn/modules/_functions.py#L202-L210) on the *forwards* pass, which then triggers the behavior seen in this ticket, and we lose the subclass.

I can re-open a new issue with that behavior if it's better place to pursue the issue directly.  Certainly that one "feels" much more starkly like a clear bug than the behavior in this issue.

**Comment by albanD** [comment]
Ho interesting.
We could update that Function there with `res = (inp.view_as(inp) for inp in args)`.
That would allow the view_as to go through the `__torch_function__` call and for you to generate a proper subclass.

**Comment by nelhage** [comment]
That would definitely solve my problem! 

And for the current issue I think a doc note somewhere explaining the `torch.autograd.Function`/subclass interaction and offering the `view_as` workaround for user-authored functions if necessary would certainly be a good improvement.

**Comment by albanD** [comment]
>  think a doc note somewhere explaining the torch.autograd.Function/subclass interaction

I don't think we have such doc because they are not expected to work well. But I'm curious which part of the doc you would expect to find that information?

> That would definitely solve my problem!

I think we should update the custom Function used by the hook to use view_as directly. Do you want to send a PR for that?

**Comment by ezyang** [comment]
@albanD arguably we should do this in the AD engine directly. Since it's custom Python autograd functions, it doesn't seem like a big problem to route through higher level Python handler here.

**Comment by albanD** [comment]
The reason it is not done right now is that this logic lives in the shared code for python Function and c++ Function.
But we could also have a special case and route that back to python yes.

**Comment by nelhage** [comment]
> But I'm curious which part of the doc you would expect to find that information?

The places I looked while trying to figure out this were documented were [the docs on custom subclasses](https://pytorch.org/docs/stable/notes/extending.html#subclassing-torch-tensor) and [the docs on autograd.Function](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd), both on the "Extending PyTorch" page.
